### PR TITLE
Gossip and better download (+Arch instructions)

### PR DESCRIPTION
**Pull request by cytadela8**
Hi I made new download code and implemented gossip algorithm. I'm in the middle of debugging and can't seam to tell why current code causes problems with market test. I also had problems with fork_test (I have mine and better version of it) and lightning test that were caused by timing issues. Does were fixed by extending the delays. Delays of course will have to be fine-tuned.

This closes #154

### CODE REVIEW

File: apps/ae_core/src/consensus/push_block.erl [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    {noreply, stop};
+handle_cast({known,Block}, {go, N, Peers}) ->
+    {noreply, {go, N-1, Peers}};
+handle_cast({unknown, Block}, {go, N, Peers}) ->
+    push(Block, gossip_stop_count()-N+1),
+    {noreply, {go, gossip_stop_count(), Peers}};
+handle_cast({push,Block}, {go, N, [Peer | Peers]}) ->
+    spawn(fun() ->
+                  Resp = push_to_peer(Block, Peer), 
+                  erlang:display(Resp),
```
Comment (sennui) [inline apps/ae_core/src/consensus/push_block.erl]: why not lager

File: apps/ae_core/src/consensus/push_block.erl [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+stop() -> gen_server:cast(?MODULE, stop).
+
+gossip_stop_count() ->
+    {ok, N} = application:get_env(ae_core, push_block_gossip_stop_count),
+    N.
+
+status() ->
+    gen_server:call(?MODULE, status).
+
+remote_peer(Transaction, Peer) ->
```
Comment (sennui) [inline apps/ae_core/src/consensus/push_block.erl]: its used now in more than one place - we need to extract it to some sort of utils

File: apps/ae_http/src/ext_handler.erl [hunk 3]
```diff
@@ -32,16 +31,25 @@ doit({pubkey}) -> {ok, keys:pubkey()};
 %doit({height}) -> {ok, block_tree:height()};
 %doit({total_coins}) -> {ok, block_tree:total_coins()};
 doit({give_block, SignedBlock}) -> 
-    %true = block:height(SignedBlock) < api:height() + 2,
-    io:fwrite("received block\n"),
     block_absorber:enqueue(SignedBlock),
     {ok, 0};
+doit({give_block_new, SignedBlock}) ->
```
Comment (sennui) [inline apps/ae_http/src/ext_handler.erl]: give_new_block ?

File: apps/ae_http/src/ext_handler.erl [hunk 4]
```diff
@@ -32,16 +31,25 @@ doit({pubkey}) -> {ok, keys:pubkey()};
...
 doit({give_block, SignedBlock}) -> 
-    %true = block:height(SignedBlock) < api:height() + 2,
-    io:fwrite("received block\n"),
     block_absorber:enqueue(SignedBlock),
     {ok, 0};
+doit({give_block_new, SignedBlock}) -> 
+    %true = block:height(SignedBlock) < api:height() + 2, %removed becouse we may get blocks faster then we can process them
+    case block_hashes:check(block:hash(SignedBlock)) of
+        true ->
+            {ok, "known"};
```
Comment (sennui) [inline apps/ae_http/src/ext_handler.erl]: please don't use strings. It's perfect place to use atoms.

File: tests/test_market.py [hunk 5]
```diff
@@ -39,12 +39,12 @@ def test_market(self):
         self.request('trade', DEV_1_INT, [3000, 1, 1, 2, 20, [127,0,0,1], 3030], sleep=0.1)
```
Comment (sennui) [inline tests/test_market.py]: can you increase to 1 sec all sleeps, just to eliminate timing issue

File: apps/ae_core/src/consensus/push_block.erl [hunk 6]
```diff
@@ -0,0 +1,82 @@
+-module(push_block).
```
Comment (sennui) [inline apps/ae_core/src/consensus/push_block.erl]: please use standard template for gen_serv... the lack of spaces makes it really difficult to read

File: apps/ae_core/src/consensus/push_block.erl [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+handle_cast(_, {go, 0, _}) ->
+    {noreply, stop};
+handle_cast({known,Block}, {go, 1, _}) ->
+    {noreply, stop};
+handle_cast({known,Block}, {go, N, Peers}) ->
+    {noreply, {go, N-1, Peers}};
+handle_cast({unknown, Block}, {go, N, Peers}) ->
+    push(Block, gossip_stop_count()-N+1),
+    {noreply, {go, gossip_stop_count(), Peers}};
+handle_cast({push,Block}, {go, N, [Peer | Peers]}) ->
```
Comment (sennui) [inline apps/ae_core/src/consensus/push_block.erl]: the peer list is kind of dynamic (for push/2 API)?
If one calls push/2 the peers are removed and then need to be re-instantiated from peers:all? How to do it? I think I see it now -- you just have the gen_srv for one round.

It seems to be correct ^^ and you stop the gen_srv from within 'push' spawned fun.

This is a bit too expensive to first, instantiate the gen_srv and then spawn internal call.
But maybe its OK, given how heavy is mining blocks and actually making a remote call.

File: apps/ae_core/src/consensus/download_blocks.erl [hunk 8]
```diff
@@ -8,75 +8,103 @@
...
 sync_all([Peer|T], Height) ->
     spawn(fun() ->
-		  sync(Peer, Height)
-	  end),
+                  sync(Peer, Height)
+          end),
     sync_all(T, Height).
 
-
 sync(Peer, MyHeight) ->
```
Comment (sennui) [inline apps/ae_core/src/consensus/download_blocks.erl]: you need to walk me through those changes

File: apps/ae_core/src/consensus/chain/block_absorber.erl [hunk 9]
```diff
@@ -23,6 +24,13 @@ handle_cast(garbage, X) ->
     {noreply, X};
 handle_cast({doit, BP}, X) ->
     absorb_internal(BP),
+    {noreply, X};
+handle_cast({doit_and_push, BP}, X) ->
```
Comment (sennui) [inline apps/ae_core/src/consensus/chain/block_absorber.erl]: this is the force push, when we mine a block?

**Comment by cytadela8** [comment]
Tried to fix stuff, nicer style, longer delays

**Comment by cytadela8** [comment]
Tried to fix stuff, nicer style, longer delays

**Comment by cytadela8** [comment]
`MarketTest` still failes.

File: tests/test_lightning.py [hunk 10]
```diff
@@ -11,6 +11,8 @@
 
 class LightningTest(ApiUser):
     def test_payments(self):
+        import time
+        time.sleep(5)  # for some reason test_fork makes stuff run in backgroud for quite some time
```
Comment (zp-sd) [inline tests/test_lightning.py]: 5 secs? 🤔

File: tests/test_market.py [hunk 11]
```diff
@@ -3,54 +3,54 @@
...
+        self.request('oracle_bet', DEV_1_INT, [2, 1, 269], sleep=1)
+        self.request('mine_block', DEV_3_INT, [11, 1], sleep=1)
+        self.request('oracle_close', DEV_1_INT, [2], sleep=1)
+        self.request('sync', DEV_1_INT, [[127, 0, 0, 1], 3030], sleep=1)
+        self.request('sync', DEV_1_INT, [[127, 0, 0, 1], 3020], sleep=1)
+        self.request('settle_bets', DEV_3_INT, [], sleep=1)
+        self.request('oracle_shares', DEV_1_INT, [1], sleep=1)
+        self.request('oracle_shares', DEV_1_INT, [2], sleep=1)
+        #self.request('oracle_unmatched', DEV_1_INT, [1, 1], sleep=1)
+        #self.request('oracle_unmatched', DEV_1_INT, [2, 1], sleep=1)
```
Comment (zp-sd) [inline tests/test_market.py]: ✂️  commented out code

File: apps/ae_core/src/consensus/download_blocks.erl [hunk 12]
```diff
@@ -8,75 +8,101 @@
...
+    TopResp = talker:talk({top}, Peer),
+    case TopResp of
+        {error, failed_connect} -> 
+            io:fwrite("failed connect"),
+            ok;
+        {ok, TopBlock, Height}  ->
+            %io:fwrite("got topblock\n"),
+            {ok, DBB} = application:get_env(ae_core, download_blocks_batch),
+	    HH = MyHeight + DBB,
+	    if
```
Comment (zp-sd) [inline apps/ae_core/src/consensus/download_blocks.erl]: if -> case would be more readable to me

File: apps/ae_core/src/consensus/download_blocks.erl [hunk 13]
```diff
@@ -8,75 +8,101 @@
...
+		    %io:fwrite("HH < Height\n"),
+                    {ok, Sizecap} = application:get_env(ae_core, download_blocks_sizecap),
+                    {ok, Blocks} = talker:talk({block_sizecap, HH, Sizecap}, Peer),
+                    trade_blocks(Peer, Blocks);
+                true ->
+                    trade_blocks(Peer, [TopBlock])
+            end,
+            get_txs(Peer),
+	    trade_peers(Peer);
+            X -> io:fwrite(X)
```
Comment (zp-sd) [inline apps/ae_core/src/consensus/download_blocks.erl]: `io:fwrite` -> `lager`

File: tests/test_market.py (hunk 11 above)
Comment (cytadela8) [inline tests/test_market.py]: not mine code

**Comment by cytadela8** [comment]
There are conflicts again and `test_market.py` makes problems.

**Comment by cytadela8** [comment]
I close this pull request, due to being to complex. 

The parts of this PR:
- Arch instructions is in #298 
- `block_sizecap` external api is in #299 
- `download_blocks.erl`using `block_sizecap` and fixing insignificant fork synchronization problem is dropped as will be replaced by planned new peers and sync code discussed in #292  
- gossip code (`push_block.erl`) is in #297

**Pull request closed without merging**